### DISCUSSION TITLE
Monorepo with Yarn workspaces fails to find modules

### DISCUSSION THREAD

**Discussion by heysailor** [vercel/vercel#5132] [Help] [answered]
I have a problem deploying a monorepo project to Vercel. I have a classic Yarn workspaces structure with shared modules, as well as multiple Vercel projects. The Vercel projects contain serverless functions in an `api` directory.

```
- projects/pubsub/package.json
- projects/pubsub/api/publish/[topicId].ts
- shared/[multiple packages]
package.json
```

I've set up the sparkly new vercel monorepo capability by setting the project root to `projects/pubsub`. Yarn installs dependencies fine locally, honouring my workspace packages. Working locally with `vercel dev` all works fine:

```
> Ready! Available at http://localhost:4000
> [debug] [2020-09-01T06:40:13.360Z] POST /api/publish/accounts.user.created
> [debug] [2020-09-01T06:40:13.360Z] Reading `package.json` file
> [debug] [2020-09-01T06:40:13.360Z] Reading `vercel.json` file
> [debug] [2020-09-01T06:40:13.361Z] No `vercel.json` file present
[if /projects/pubsub/package.json is absent, "No `package.json` file present" appears here, so its absence tells me it is reading the file]
> [debug] [2020-09-01T06:40:13.361Z] Locating files /Volumes/Drake/dev/climbingzone/climbingzone/projects/pubsub
> [debug] [2020-09-01T06:40:13.362Z] Ignoring /Volumes/Drake/dev/climbingzone/climbingzone/projects/pubsub/.DS_Store
> [debug] [2020-09-01T06:40:13.362Z] Ignoring /Volumes/Drake/dev/climbingzone/climbingzone/projects/pubsub/.vercel
> [debug] [2020-09-01T06:40:13.362Z] Ignoring /Volumes/Drake/dev/climbingzone/climbingzone/projects/pubsub/node_modules
> [debug] [2020-09-01T06:40:13.362Z] Locating files /Volumes/Drake/dev/climbingzone/climbingzone/projects/pubsub: 0.854ms
> [debug] [2020-09-01T06:40:14.061Z] Proxying to "@vercel/node" dev server (port=57227, pid=70516)
accounts.user.created topic event published with ID 1526894198759698 <-- logged on success by serverless function
```

But on deployment on Vercel, calling the api triggers the error:

```
[POST] /api/publish/accounts.user.created
14:50:04:04
2020-09-01T06:50:04.587Z	undefined	ERROR	Cannot find module 'micro-secret'
Require stack:
- /var/task/api/publish/[topicId].ts
- /var/task/___now_launcher.js
- /var/runtime/UserFunction.js
- /var/runtime/index.js
2020-09-01T06:50:04.587Z	undefined	ERROR	Did you forget to add it to "dependencies" in `package.json`?
RequestId: 5e5b817f-2e51-4560-9f6a-a7d8f6ba7407 Error: Runtime exited with error: exit status 1
Runtime.ExitError
```

... `micro-secret` is a module on npm. The `projects/pubsub/package.json` file lists micro-secret as a dependency - and I've tried putting a `package.json` adjacent to `[topicId].ts` too. I have enabled the `Include source files outside of the Root Directory in the Build Step` option for the project.

I can cause the same error to occur with `vercel dev` if the root `node_modules` directory is missing.

What am I missing? I'm guessing _only_ the monorepo root `package.json` is being read and its dependencies installed. How do I ensure the dependencies of `projects/pubsub/package.json` are correctly installed and made available to the serverless functions? Or is this a bug in the new monorepo feature?

**Comment by heysailor** [comment, marked as answer]
[Here's a Yarn workspaces monorepo with the issue.](https://github.com/climbingzone/vercel-monorepo)

- this has one project in projects/name
- its serverless function at `/api/random` should produce a random name from the 'unique-names-generator' external node module

The repo works fine locally with vercel dev after yarn install and linking the project with vercel.

After importing the Github repository to Vercel and setting the project root as projects/name, and including source files outside the project root, the vercel deployment URL is [https://vercel-monorepo.climbingzone.vercel.app/api/random](https://vercel-monorepo.climbingzone.vercel.app/api/random) - currently returning a `502 NO_RESPONSE_FROM_FUNCTION` error. 

The error reported in the functions tab of the dashboard is:

```
2020-09-01T14:02:11.731Z	undefined	ERROR	Cannot find module 'unique-names-generator'
Require stack:
- /var/task/api/random/index.ts
- /var/task/___now_launcher.js
- /var/runtime/UserFunction.js
- /var/runtime/index.js
2020-09-01T14:02:11.732Z	undefined	ERROR	Did you forget to add it to "dependencies" in `package.json`?
RequestId: ff5aef8c-c134-4a54-8d79-3881df2d1d41 Error: Runtime exited with error: exit status 1
Runtime.ExitError
```

**Reply by corydeppen** [reply]
A fix that was recently deployed and mentioned here may help.

https://github.com/vercel/next.js/issues/16667#issuecomment-687624723

**Reply by heysailor** [reply]
Excellent. That fix makes the repro above work. Thanks all.
